Gatsby Plugin CLI

moonmeister opened the pull request:
See: https://github.com/gatsbyjs/rfcs/pull/13 for discussion of feature and implementation details.

UI Tasks: 
- [x] Add "add" Command to CLI
- [x] Add "remove" Command to CLI
- [x] Add "search" Command to CLI
- [x] Add "config" Command to CLI
- [x] Add Confirmation for add/remove actions before executing
- [x] On completion of add/remove log completion and plugin docs to console
- [ ] ~~On completion of config log new config to console and wait for confirmation before writing~~

Implementation Tasks:
- [x] "add <plugin..>" add to package.json and install from npm
- [x] "remove <plugin..>" remove from package.json and remove from node_modules
- [ ] ~~"search <query>" command~~
- [ ] ~~extend gatsby API with 'onConfigPlugin' to add config to gatsby-config.js **in progress:** @moonmeister see: https://github.com/gatsbyjs/gatsby/tree/config-api~~
- [ ] ~~"config <plugin..>" command~~
- [ ] ~~add hooks for "add" to config added plugins~~
- [ ] ~~"remove" remove config from gatsby-config~~
- [ ] ~~auto search into add/remove/config~~
- [x] make library for code shared with "new" CLI (yarn choice and spawn)

Testing Tasks:
- [ ] Test all the things

Docs/Community Tasks:
- [ ] Document CLI
- [ ] ~~Document API Changes~~
- [ ] ~~Update plugins in the Gatsby repo to be compatible~~
- [ ] ~~Tutorial for plugin creators to make their plugins compatible with the configuration work flow~~
- [ ] Blog post announcing this sweet sweet feature!

Happy to take PRs against this branch to complete tasks. Please comment if you're going to take a task so we know!


UPDATE [18-03-2019]: Crossed out things no longer being done on this PR. It's not that we don't like these ideas. We're just simplifying this PR.

moonmeister commented:
I'm going to break the API extension for configuring a plugin into a separate PR and will link when I open that PR. Reason being is I think it's functionality can go beyond this CLI extension. It could theoretically also be added to starters and themes. Then it could be triggered when any of things are installed.

@jlengstorf I'm starting to wonder if this api needs to be it's own RFC...Thoughts?

The code is available here: https://github.com/gatsbyjs/gatsby/tree/config-api

jlengstorf commented:
@moonmeister it seems to me that you can build it out for this use case, but make sure it's extendable for others. No need to solve all the problems at once. 😄

wardpeet commented:
@moonmeister any chance you could pick this one up again? We would be ok to go forward with add & remove in this PR 👍

moonmeister commented:
@wardpeet yeah, I've been thinking about it lately. Happy to finish it. Can you give me some feedback on the code as is? Add/remove are implemented in this currently. They probably need some finishing touches though.

moonmeister commented on packages/gatsby/package.json:
```diff
@@ -70,6 +70,7 @@
     "graphql-tools": "^3.0.4",
     "graphql-type-json": "^0.2.1",
     "hash-mod": "^0.0.5",
+    "inquirer": "^6.2.1",
```
yeah, I'll check out enquirer. As far as it's usefulness, I agree, the current feature set is pretty pointless. To implement the config stuff we'd need a new api, I've started playing with this but it sounded like we wanted to get this feature in. I say we get this first step done and not worry about announcing it or making a big deal about it. 

We can continue working on a variety of features around this plugin CLI and once the config/search/etc features are done and this becomes more useful, we can make doc changes and a blog post.

moonmeister commented:
@wardpeet @dschau where we at an this. Related config API work is moving forward in other PRs. Given our conversation about breaking up large projects it seems reasonable to merge this partial one. I will say the code probably needs a little more work and I may have some changes to merge in from my configapi branch that can happen now. Thoughts?

moonmeister commented:
@sidharthachatterjee Okay, updated the code from my other branch and rebased. I still need to figure out the Yarn vs NPM using the new "standard" way. Are there docs or a PR you can point me to get me on my way?

DSchau commented:
Hey Alex!

First of all--just wanted to convey how much we appreciate this PR and our appreciation of the excellent work you've done thus far.

We think the work you've done will be foundational in our future work to make doing "Gatsby things" more automated and easier--and we see this type of work as crucial in making Gatsby more friendly and more accessible to developers of all skill levels. This is something we care about deeply, and we love that you've been working on making certain Gatsby actions easier to do and automated.

That being said--I don't think we want to merge this in its current state. Why?

The PR is clearly foundational, and sets the stage for more impactful and valuable work in the future. For instance, it sets up the ability to do something like `gatsby plugin add gatsby-source-contentful`, which currently will install the plugin with npm/yarn. However -- in its current state, that's not really a valuable outcome, and just adds some indirection and confusion to the developer (e.g. why would I run that vs. just `yarn add`). I _know_ you've worked on other aspects of it (e.g. parsing the AST to add plugin(s) in #12822), but those additional, more useful tasks are blocked by us as we further investigate this space and formulate how we best want to establish building blocks to support these kinds of features--for example a plugin validation API and other Gatsby Core API-level implementations. Additionally, the separate, more meaningful features are quite a bit more complex and we want to ensure that we establish the right level of helpers/API to ensure we can roll these out as a first-class CLI integration without brittleness and flakiness.

From this basis, let's close this PR out as it's not in a mergeable state currently--which is every bit our fault for not giving more direction before and as you began the implementation of the PR.

Again -- we appreciate your work here more than you know; thanks for sticking around and being diligent about keeping this up to date! I fully anticipate we'll revisit this type of work in the future, and we'd love your assistance, insight, and expertise when we do so. We'll be in touch!

jmolivas commented:
Hey @DSchau Symfony had something called Flex

https://flex.symfony.com/

> It is based on Symfony Recipes, which are a set of automated instructions to integrate third-party packages into Symfony applications.

Those are configuration recipes that got added via a post-script once you add you package

I added a comment since this issue and link appeared on a recent conversation on discord
